go/consensus/tendermint/apps/governance/governance.go

```diff
@@ -382,24 +385,95 @@ func (app *governanceApplication) closeProposal(
+	validatorVoteShares := make(map[stakingAPI.Address]map[governance.Vote]quantity.Quantity)
+	for validator := range validatorEntitiesPool {
+		validatorVoteShares[validator] = make(map[governance.Vote]quantity.Quantity)
+	}
+
+	// Tally the validator votes.
 	for _, vote := range votes {
-		escrow, ok := validatorEntitiesEscrow[vote.Voter]
+		escrow, ok := validatorEntitiesPool[vote.Voter]
 		if !ok {
-			// Voter not in current validator set - invalid vote.
-			proposal.InvalidVotes++
+			// Skip non-validator votes.
 			continue
 		}
+		validatorVotes[vote.Voter] = &vote.Vote
+		if err = addShares(validatorVoteShares[vote.Voter], vote.Vote, escrow.TotalShares); err != nil {
+			return fmt.Errorf("failed to add shares: %w", err)
+		}
+	}
 
-		currentVotes := proposal.Results[vote.Vote]
-		newVotes := escrow.Clone()
-		if err := newVotes.Add(&currentVotes); err != nil {
-			return fmt.Errorf("failed to add votes: %w", err)
+	// Tally delegator votes.
+	for _, vote := range votes {
+		// Fetch outgoing delegations.
+		delegations, err := stakingState.DelegationsFor(ctx, vote.Voter)
+		if err != nil {
+			ctx.Logger().Error("failed to fetch delegations for",
+				"delegator", vote.Voter,
+				"err", err,
+			)
+			return fmt.Errorf("failed to fetch delegations: %w", err)
+		}
+		var delegationToValidator bool
+		for to, delegation := range delegations {
+			if _, ok := validatorEntitiesPool[to]; !ok {
+				continue
+			}
+			delegationToValidator = true
+			validatorVote := validatorVotes[to]
+			if validatorVote == &vote.Vote {
+				// Vote matches the delegated validator vote.
+				continue
+			}
+
+			// Deduct shares from the validators shares.
+			if validatorVote != nil {
+				if err := subShares(validatorVoteShares[to], *validatorVote, delegation.Shares); err != nil {
+					return fmt.Errorf("failed to sub votes: %w", err)
+				}
+			}
+
+			// Add shares to the voters vote.
+			if err := addShares(validatorVoteShares[to], vote.Vote, delegation.Shares); err != nil {
+				return fmt.Errorf("failed to add votes: %w", err)
+			}
+		}
+		if !delegationToValidator {
+			proposal.InvalidVotes++
+		}
+	}
+
+	// Convert vote in shares into stake.
+	for validator, votes := range validatorVoteShares {
+		validatorPool, ok := validatorEntitiesPool[validator]
+		if !ok {
+			panic("should not happen")
+		}
+		for vote, shares := range votes {
+			// Compute stake from shares.
+			escrow, err := validatorPool.StakeForShares(shares.Clone())
+			if err != nil {
+				ctx.Logger().Error("failed to compute stake from shares for",
+					"share_pool", validatorPool,
+					"validator", validator,
+					"err", err,
+				)
+				return fmt.Errorf("failed to compute stake from shares: %w", err)
+
+			}
+
+			// Add stake to vote.
+			currentVotes := proposal.Results[vote]
+			if err := currentVotes.Add(escrow); err != nil {
+				return fmt.Errorf("failed to add votes: %w", err)
+			}
+			proposal.Results[vote] = currentVotes
+			// TODO: also add total stake here.
```

Comment:
What is this TODO about?

Reply:
Outdated, I thought about moving the total voting stake computation here, but that would be wrong.